### PR TITLE
possible dependent config leak

### PR DESCRIPTION
**Pull request by ramaraochavali**
Built on top of https://github.com/istio/istio/pull/39853.

One possible leak I think is when a configUpdate happens, only one of dependent config change, might trigger a cache key to evicted but the key might exist in other config reference. This PR cleans up key on evict from dependent config list. 

Main changes are in `onEvict` - always clean index in `onEvict`.

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/model/xds_cache.go [hunk 1]
```diff
@@ -185,19 +188,22 @@ func (l *lruCache) recordDependentConfigSize() {
 }
 
 // This is the callback passed to LRU, it will be called whenever a key is removed.
-func (l *lruCache) evict(k interface{}, v interface{}) {
+func (l *lruCache) onEvict(k interface{}, v interface{}) {
 	if features.EnableXDSCacheMetrics {
-		xdsCacheEvictions.Increment()
-	}
-
-	if !l.evicted {
```
Comment (sschepens) [inline pilot/pkg/model/xds_cache.go:193]: yes, this seems like a clear leak, the tests I had on another PR should've faild with this change

Comment (sschepens) [reply pilot/pkg/model/xds_cache.go:193]: i see that https://github.com/istio/istio/pull/39688 has been reverted, which changed the behaviour of the cache and made my tests invalid, I could now probably merge them.

**Review by howardjohn — APPROVED** [review]
lgtm, can you fix the conflict though?

**Review by adiprerepa — APPROVED** [review]

File: pilot/pkg/model/xds_cache.go [hunk 2]
```diff
@@ -341,9 +347,9 @@ func (l *lruCache) Clear(configs map[ConfigKey]struct{}) {
 	defer l.mu.Unlock()
 	l.token = CacheToken(time.Now().UnixNano())
 	// not to call evict on keys that are removed actively
-	l.evicted = false
+	l.evictedOnClear = true
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/xds_cache.go:350]: remove the comment

**Pull request merged**